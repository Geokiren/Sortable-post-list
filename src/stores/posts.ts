import { defineStore } from 'pinia';
import type { IPosts } from '@/types/posts';
import type { IHistory } from '@/types/history';
import { getPosts } from '@/services/postsService';
import type { AlertsType, IAlert } from '@/types/alerts';

export const usePostsStore = defineStore('posts', {
  state: () => ({
    posts: [] as IPosts,
    history: [] as IHistory,
    apiBaseUrl: import.meta.env.VITE_API_BASE_URL as string,
    pagination: {
      page: 1,
      limit: 5
    } as { page: number; limit: number },
    alerts: [] as AlertsType
  }),

  getters: {
    historyReversed(state): IHistory {
      return state.history.reverse();
    }
  },

  actions: {
    async fetchPosts() {
      if (!this.apiBaseUrl) {
        this.alerts.push({
          id: crypto?.randomUUID ? crypto.randomUUID() : Date.now(),
          type: 'error',
          message: 'API URL not set'
        });

        return;
      }

      try {
        const data = await getPosts(this.pagination);

        this.posts.push(...data);
      } catch (error) {
        this.addAlert('Could not fetch posts. Please, try again later.', 'error');
      }
    },

    addAlert(message: string, type: IAlert['type']) {
      this.alerts.push({
        id: crypto?.randomUUID ? crypto.randomUUID() : Date.now(),
        type,
        message
      });
    },

    swapIndexes(from: number, to: number) {
      const fromItem = this.posts[from];

      this.history.push({
        id: crypto?.randomUUID ? crypto.randomUUID() : Date.now(),
        message: `Moved Post ${fromItem.id} from index ${from} to index ${to}`,
        posts: this.posts.slice()
      });

      this.posts[from] = this.posts[to];
      this.posts[to] = fromItem;
    },

    timeTravel(index: number) {
      this.posts.splice(0, this.posts.length, ...this.historyReversed[index].posts);
      this.history.length = this.historyReversed.length - 1 - index;
    }
  }
});
