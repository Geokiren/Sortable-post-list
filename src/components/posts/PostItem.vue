<template>
  <div class="post-item card p-2 md:p-4 h-24 lg:h-32 mb-4 md:mb-6 flex justify-between items-center text-text-primary text-xl font-semibold" data-test="post-item">
    <div>Post {{ id }}</div>

    <div class="flex flex-col h-full" :class="buttonsContainerClass">
      <button v-if="showUp" class="caret" @click="getUp">
        <CaretSvg classes="w-5 h-5 fill-secondary" />
      </button>

      <button v-if="showDown" class="caret rotate-180" @click="getDown">
        <CaretSvg classes="w-5 h-5 fill-secondary" />
      </button>
    </div>
  </div>
</template>

<script setup lang="ts">
  import { computed } from 'vue';
  import CaretSvg from '../shared/CaretSvg.vue';
  import { usePostsStore } from '@/stores/posts';

  const props = defineProps<{
    id: number,
    index: number,
    size: number,
  }>();

  const { swapIndexes }: { swapIndexes: (from: number, to: number) => void } = usePostsStore();

  const getDown = () => {
    if (typeof props.index === 'number') {
      swapIndexes(props.index, props.index + 1);
    }
  }

  const getUp = () => {
    if (typeof props.index === 'number') {
      swapIndexes(props.index, props.index - 1);
    }
  }

  const showUp = computed(() => typeof props.index === 'number' && props.index > 0);
  const showDown = computed(() => typeof props.index === 'number' && props.index < props.size - 1);
  const buttonsContainerClass = computed(() => (showUp.value && showDown.value) ? 'justify-between' : 'justify-center');
</script>