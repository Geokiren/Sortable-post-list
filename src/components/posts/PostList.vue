<template>
  <section id="post-list" class="flex flex-col gap-4 w-full md:w-1/2">
    <h2 class="text-2xl md:text-3xl font-bold text-post-title mb-2 sm:mb-5 tracking-wide">Sortable Post List</h2>
    
    <TransitionGroup name="post-list" tag="div" class="transition-group-posts" data-test="transition-group-posts">
      <PostItem v-for="(post, index) in posts" :key="post.id" :id="post.id" :index="index" :size="posts.length" />
    </TransitionGroup>
  </section>
</template>

<script setup lang="ts">
  import { usePostsStore } from '@/stores/posts';
  import type { IPosts } from '@/types/posts';
  import { onMounted } from 'vue';
  import PostItem from './PostItem.vue';

  const { posts, fetchPosts }: { posts: IPosts, fetchPosts: () => void } = usePostsStore();

  onMounted(() => {
    fetchPosts();
  });
</script>

<style scoped>
  .post-list-move {
    transition: all 0.5s;
  }
</style>