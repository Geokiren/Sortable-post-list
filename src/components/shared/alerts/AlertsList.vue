<template>
  <div id="alerts -list" v-if="alerts.length" class="alerts flex flex-col gap-1 justify-center items-end p-4" >
    <AlertMessage v-for="alert in alerts" :key="alert.id" :message="alert.message" :variant="alert.type" />
  </div>
</template>

<script setup lang="ts">
  import AlertMessage from './AlertMessage.vue';
  import { usePostsStore } from '@/stores/posts';
  import type { AlertsType } from '@/types/alerts';

  const { alerts }: { alerts: AlertsType } = usePostsStore();
</script>