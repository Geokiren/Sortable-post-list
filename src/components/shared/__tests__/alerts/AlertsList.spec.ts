import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import AlertsList from '../../alerts/AlertsList.vue';
import { usePostsStore } from '@/stores/posts';
import { createPinia } from 'pinia';
import type { AlertsType } from '@/types/alerts';

describe('AlertsList', () => {
  const mockData: AlertsType = [
    { id: 1, type: 'error', message: 'Error 1' },
    { id: 2, type: 'warning', message: 'Warning 1' }
  ];

  it('Renders properly when no store data', () => {
    const store = usePostsStore(createPinia());

    const wrapper = mount(AlertsList, { store });

    expect(wrapper.html()).toBe('<!--v-if-->');
  });

  it('Renders properly with the correct elements and text when having store data', () => {
    const store = usePostsStore(createPinia());

    const alerts = (store.$state.alerts = mockData);

    const wrapper = mount(AlertsList, {
      store: { alerts }
    });

    const alertElements = wrapper.findAll('[data-test="alert-message"]');

    expect(alertElements.length).toBe(2);
    expect(alertElements[0].text()).toBe('Error 1');
    expect(alertElements[1].text()).toBe('Warning 1');
  });

  it('Matches snapshot', () => {
    const store = usePostsStore(createPinia());

    const alerts = (store.$state.alerts = mockData);

    const wrapper = mount(AlertsList, { store: { alerts } });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
