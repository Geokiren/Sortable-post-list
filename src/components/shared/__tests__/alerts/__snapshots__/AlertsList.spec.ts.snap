// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlertsList > Matches snapshot 1`] = `
"<div id="alerts -list" class="alerts flex flex-col gap-1 justify-center items-end p-4">
  <div data-v-e382e1fe="" class="alert flex items-center px-4 py-2 rounded-xl relative z-10 undefined bg-red-500 text-red-100 w-full" data-test="alert-message">
    <div data-v-e382e1fe="" class="flex justify-center w-11/12">
      <div data-v-e382e1fe="">Error 1</div>
    </div>
    <div data-v-e382e1fe="" class="flex justify-end w-1/12"><svg data-v-e382e1fe="" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" class="fill-red-100 w-5 h-5 rotate-90 cursor-pointer transition-all duration-500" data-test="caret-svg">
        <path d="M201.4 137.4c12.5-12.5 32.8-12.5 45.3 0l160 160c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 205.3 86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l160-160z"></path>
      </svg></div>
  </div>
  <div data-v-e382e1fe="" class="alert flex items-center px-4 py-2 rounded-xl relative z-10 undefined bg-yellow-500 text-yellow-100 w-full" data-test="alert-message">
    <div data-v-e382e1fe="" class="flex justify-center w-11/12">
      <div data-v-e382e1fe="">Warning 1</div>
    </div>
    <div data-v-e382e1fe="" class="flex justify-end w-1/12"><svg data-v-e382e1fe="" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" class="fill-yellow-100 w-5 h-5 rotate-90 cursor-pointer transition-all duration-500" data-test="caret-svg">
        <path d="M201.4 137.4c12.5-12.5 32.8-12.5 45.3 0l160 160c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 205.3 86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l160-160z"></path>
      </svg></div>
  </div>
</div>"
`;
